kentik_api_library/kentik_api/api_connection/api_connector.py

```diff
@@ -37,15 +57,29 @@ def send(self, api_call: APICall, payload: Optional[Dict[str, Any]] = None) -> A
             response = requests.delete(url, auth=self._auth, headers=self.DEFAULT_HEADERS, json=payload)
         else:
             raise ValueError(f"Improper API call method: {api_call.method}")
+        return response
 
-        self._validate_response(response)
+    def _get_api_query_url(self, url_path: str) -> str:
+        return self._api_url + url_path
 
-        return APICallResponse(response.status_code, response.text)
+    @staticmethod
+    def _handle_error(response: requests.Response) -> None:
+        if response.status_code >= 400:
+            raise new_api_error(response.text, response.status_code)
 
-    def _get_api_query_url(self, api_method: str) -> str:
-        return self._api_url + api_method
 
-    def _validate_response(self, response: requests.Response) -> None:
-        # http error handling can be implemented here eg with exceptions, Expected[return, error] or error codes
-        if response.status_code >= 400:
-            self._logger.error("code: %d, body: %s", response.status_code, response.text)
+def new_api_error(message: str, http_status_code: int) -> KentikAPIError:
+    common_errors = {
+        HTTPStatus.BAD_REQUEST.value: BadRequestError,
+        HTTPStatus.UNAUTHORIZED.value: AuthError,
+        HTTPStatus.FORBIDDEN.value: AuthError,
+        HTTPStatus.NOT_FOUND.value: NotFoundError,
+        HTTPStatus.TOO_MANY_REQUESTS.value: RateLimitExceededError,
+        HTTPStatus.SERVICE_UNAVAILABLE.value: UnavailabilityError,
+        HTTPStatus.GATEWAY_TIMEOUT.value: UnavailabilityError,
+    }
+    try:
+        error = common_errors[http_status_code](message)
+    except KeyError:
+        return KentikAPIError(message)
```

Comment:
I think we should preserve the http status code besides the message, so that developers using the library can react to any specific http response scenarios they want; lets not take this information away from them.

Reply:
Adding http_status_code field to KentikAPIError would bind library closer to implementation of transport layer and increase API surface. My idea here was to add new errors (such as NotFoundError) for each new HTTP status code that needs to be distinguished by users. We would add new errors on new feature requests reported by users.
Decoupling from transport layer would allow to switch it to other protocols like gRPC without breaking compatibility, but chances for that are probably low.
It might be too strict approach, what do you think @mateuszmidor @lukaa12 ?

Reply:
I like the idea of switching to compiled protocol; someone has already said that "Internet is running in debug mode".
And I like the idea of abstracting out the public interface from underlying protocol.
Then again, I see 3 kinds of errors we can have:
-protocol errors that currently translate directly from http codes>=400; these are probably already handled by current Kentik customers in their code
-validation errors like invalid object about to be sent
-deserialization errors - response code is 200 but body document is invalid (empty document, invalid json, json missing fields, fields of invalid types/values).
Maybe we should encapsulate "other" http/grpc errors into a separate type eg ProtocolError? gRPC also uses status codes

Reply:
Isolating library API from transport layer is a good idea. We don't have to pass http code, instead we should provide clear explanation of which http code causes which error in documentation.

Reply:
Improved error docstrings for now.

Reply:
ProtocolError added.